### DISCUSSION TITLE
exporting disassembled .text section

### DISCUSSION THREAD

**Discussion by isimsizolan** [hasherezade/pe-bear#14] [Q&A]
hi,
Can disassembled .text section exported using PE-BEAR?

**Comment by hasherezade** [comment]
hi! do you mean automatic coping/saving to a file all the disassembled code within the section? currently it is not supported, but I can add it.

**Reply by isimsizolan** [reply]
yes precisely! it will be awesome. You truly are rock star of reversing :)

**Reply by hasherezade** [reply]
@isimsizolan  - I added it. Would you like to test the new version pre-release? You can download the builds from AppVeyor server, as described in the [README](https://github.com/hasherezade/pe-bear#builds).

![export_disasm](https://user-images.githubusercontent.com/3115348/219979220-e47f58ad-2bf3-408e-a12d-a767f9769d75.png)

**Reply by isimsizolan** [reply]
I have tested it on some disarmed malwares, works like a charm! Thank you!

I have an another question for you, is there feasible way to do this exact operation (extracting disassemblies from section which contains EP) in large quantities in some kind of automation?

**Reply by hasherezade** [reply]
@isimsizolan - thanks for checking! I am glad that you like it.

Regarding exporting from multiple PEs, I guess you need some commandline tool for it, that would process i.e. a whole supplied directory with samples... I am thinking it is rather out of scope for PE-bear.

One possibility to make it in PE-bear, which comes to my mind, is, to add an action to export section with Entry Point from all the PEs currently loaded. Is this something that interests you?

**Comment by isimsizolan** [comment]
That feature could work just fine.

To test it I tried to load many malwares however random warnings pops on some pe files, and this warning can not be passed without user interaction. User has to be watching loading of all pe-files and click ok when warning pops up which is unfeasible because loading will take time.

![warning-1](https://user-images.githubusercontent.com/45662732/220423353-c8767679-75be-4403-9e35-eb408a98ef51.PNG)
![warning-2](https://user-images.githubusercontent.com/45662732/220423361-6dac7582-6d86-474f-819b-46491febaa46.PNG)

**Reply by hasherezade** [reply]
hah, ok, so you stumbled upon the problem discussed here: https://github.com/hasherezade/pe-bear/issues/11 (I had doubts whether or not should I add popups with warnings). I am gonna do something about them, possibly disable them on multiple loads.

**Reply by isimsizolan** [reply]
Hmm btw, is this aligment issue fixed on the version i tested?

**Reply by hasherezade** [reply]
@isimsizolan - yes sure it should fixed - did you encountered any related problems?
 I am more specifically talking about the whole discussion if the alerts should be showed or not. This is what I said:

>I have mixed feelings about showing popups during the load, unless it is a critical error. Some people just find them annoying. 

And here we go ;)

**Reply by hasherezade** [reply]
@isimsizolan  - I just hidden those popups. You can download the new version from AppVeyor (as soon as it finishes building). Please let me know if it is better now.

**Reply by isimsizolan** [reply]
Well i have limited experience with reversing to actually verify if it is working as intented, i just received warnings but could not verify if there is any inconsistency between pe-bear section vs actually loaded. 

About the warnings, disabling any user-interaction warning will fine as you suggested or a small log bar below to write them there will be better.

**Reply by hasherezade** [reply]
this is what I did - now the warnings are shown on the status bar (for the last loaded file, in case if you load multiple at once):

![logged_msg](https://user-images.githubusercontent.com/3115348/220430265-f2426664-b425-4098-a9ce-238b96a220f5.png)

And the same warnings, for every file, can be found displayed as a tooltip:

![messages](https://user-images.githubusercontent.com/3115348/220430409-ebed8834-a624-4b74-840e-6494ccc0ee44.png)

When the PE has this small alert icon in the corner, it means there are some warnings listed.

**Reply by isimsizolan** [reply]
Okey, i have started to load pe malware files (around 3k), well if memory survives, i'll let you know the end result. Right now no popups but ui thread blocked.

**Reply by hasherezade** [reply]
TBH, even if the memory survives, I am afraid it is all gonna be super slow. PE-bear was not intended to process big datasets like this. And in general for massive processing of samples, GUI applications like this are not the optimal solution.
IMO you should try some other, commandline tool.

**Reply by isimsizolan** [reply]
Actually it just took couple of minutes, entire set is loaded :) (total of 3095 malware samples)
So if you could just add the option to extract for all, It will work just fine.

![2135](https://user-images.githubusercontent.com/45662732/220435590-5c79f339-fd32-49e4-9725-a91070fd4186.PNG)

**Reply by hasherezade** [reply]
hmm, ok then, I can add it.

**Reply by hasherezade** [reply]
@isimsizolan - check this out - I added a new option in the main menu:
`File`->`From all loaded...`->`Export disassembly to...`

![export1](https://user-images.githubusercontent.com/3115348/220474530-050a47ad-0926-42e7-b7a9-08bd68954895.png)

**Reply by isimsizolan** [reply]
I just found time to check, you have just end my endless research for perfect tool to extract opcodes from bulk samples with altered headers. I'll highly recommend it to my colleagues with these feature set. I'll definetely mention it in my upcoming paper.  Thank you!

A minor issue was, error below. It effected very few samples. I can track them and send them to you if you request.

![12345](https://user-images.githubusercontent.com/45662732/221064136-d81b390b-c68e-42fd-a239-7dfff567d243.PNG)

**Reply by hasherezade** [reply]
I am happy to hear that it helped you! :)
and sure I will like to have a look at the set that failed to export. if you have them on some cloud drive, you can paste the link here, or contact me about it via my email: hasherezade-at-protonmail.com

**Reply by isimsizolan** [reply]
I have emailed dataset link, a list of samples which can not be exported and some other question as well 
Looking forward to hear from you.

**Reply by hasherezade** [reply]
thank you, I am gonna check soon! BTW, the changes that we discussed here are now included in the official release: https://github.com/hasherezade/pe-bear/releases/tag/v0.6.5 - but I am gonna continue to work on this feature and improve it more for the next release.

**Reply by isimsizolan** [reply]
Thank you, i saw it on twitter. I have couple of new features to suggest for pe-bear which i believe will greatly increase ability for the researchers to use machine learning models. I would like to discuss it with you soon enough. You can reach me fast over email i sent to you. Have a great weekend!

**Reply by hasherezade** [reply]
cool, I am always open for suggestions about new features. we will talk soon. have a great weekend too!
